cmd/soroban-cli/src/snapshot.rs

```diff
@@ -120,10 +127,95 @@ where
 
     Ok(())
 }
+
+/// Returns a list of events from the on-disk event store, which stores events
+/// exactly as they'd be returned by an RPC server.
+pub fn read_events(path: &std::path::PathBuf) -> Result<Vec<rpc::Event>, Error> {
+    let reader = std::fs::OpenOptions::new().read(true).open(path)?;
+    Ok(serde_json::from_reader(reader)?)
+}
+
+// Reads the existing event file, appends the new events, and writes it all to
+// disk. Note that this almost certainly isn't safe to call in parallel.
+pub fn commit_events(
+    new_events: &[events::HostEvent],
+    ledger_info: &LedgerInfo,
+    output_file: &std::path::PathBuf,
+) -> Result<(), Error> {
+    // Create the directory tree if necessary, since these are unlikely to be
+    // the first events.
+    if let Some(dir) = output_file.parent() {
+        if !dir.exists() {
+            create_dir_all(dir)?;
+        }
+    }
+
+    let mut file = std::fs::OpenOptions::new()
+        .create(true)
+        .read(true)
+        .open(output_file)?;
+    let mut events: rpc::GetEventsResponse = serde_json::from_reader(&mut file)?;
+
+    for (i, event) in new_events.iter().enumerate() {
+        let contract_event = match event {
+            events::HostEvent::Contract(e) => e,
+            events::HostEvent::Debug(_e) => todo!(),
+        };
+
+        let topics = match &contract_event.body {
+            xdr::ContractEventBody::V0(e) => &e.topics,
+        }
+        .iter()
+        .map(xdr::WriteXdr::to_xdr_base64)
+        .collect::<Result<Vec<String>, _>>()?;
+
+        let id = toid::Toid::new(
+            ledger_info.sequence_number,
+            // we should technically inject the tx order here from the ledger
+            // info, but the sandbox does one tx/op per ledger anyway, so this
+            // is a safe assumption
+            1,
+            1,
+        );
+        let cereal_event = rpc::Event {
+            event_type: "contract".to_string(),
+            id: id.to_string(),
+            // stolen from
+            // https://github.com/stellar/soroban-tools/blob/main/cmd/soroban-rpc/internal/methods/get_events.go#L264
+            paging_token: format!("{}-{:010}", id.to_paging_token(), i + 1),
+            ledger: ledger_info.sequence_number.to_string(),
+            ledger_closed_at: ledger_info.timestamp.to_string(),
+            contract_id: hex::encode(
+                contract_event
+                    .contract_id
+                    .as_ref()
+                    .unwrap_or(&xdr::Hash([0; 32])),
+            ),
+            topic: topics,
+            value: rpc::EventValue {
+                xdr: match &contract_event.body {
+                    xdr::ContractEventBody::V0(e) => &e.data,
+                }
+                .to_xdr_base64()?,
+            },
+        };
+
+        events.push(cereal_event);
+    }
+
+    let mut file = std::fs::OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .open(output_file)?;
```

Comment:
*question*
Feels a bit awkward to rewrite the entire file every time and truncate. vs append. Probably fine for v1, tho. And avoids race-condition read-inconsistencies, maybe?

Reply:
yeahhh I thought about this and basically "good enough"d it: on one hand you could dump raw xdr and just append new entries, on the other hand you can have json and readability but then you need full rewrites on appends (unless you want to be clever and inject the serialized json in the correct part of the array in the file, but it seems like we shouldn't cross that bridge if we don't need to). I guess you could also have the file contain json blobs but not be a valid json blob itself

Reply:
I dunno if race conditions are worth worrying about - who is running multiple sandbox invokes simultaneously?